kuma/api/v1/views.py

```diff
@@ -167,4 +181,12 @@ def whoami(request):
                 'large': None,
             }
         }
+
+    # Add waffle data to the data
+    data['waffle'] = {
+        'flags': {f.name: f.is_active(request) for f in all_waffle_flags},
+        'switches': {s.name: s.is_active() for s in all_waffle_switches},
+        'samples': {s.name: s.is_active() for s in all_waffle_samples},
```

Comment:
First of all, all *but* `all_waffle_flags` here could be heavily cached. Either in Redis or inside each gunicorn worker. 
This endpoint is going to see heavy traffic since it'll run for every single browser load. 
As it stands now we need to generate 3 dicts from 3 for-loops on every request. 

I don't have a better suggestion but I wanted to raise the point. Also, I'm not very experienced with django-waffle so I'm not entirely sure what I'm talking about. 

And we would certainly not want to break the Django admin. I.e. editing waffle stuff shouldn't require that we restart the servers for it to take effect. If it was hooked into the django cache framework, with a known cache key, we could write a `post_save` signal somewhere that's tied to these models.

Reply:
From a perf perspective, all of this code is already being run in the {{waffle.wafflejs()}} template tag in one of the files we import into our templates. So this is not adding any load to the server that the server isn't already handling.

Second: the only time this patch requires a server restart is when a new flag or switch or sample is added. But just adding one of those from the admin interface without deploying new code to use the flag or switch or sample won't have any impact. So, if a new flag is added, it will be useless until new code is deployed. And the deploy will cause the servers to restart. So I think we're okay here.

I guess this does mean that when deleting a flag, it would be important to first set it to false, and then delete it because existing frontend code won't see the deletion until the next deploy.

Reply:
> only time this patch requires a server restart is when a new flag or switch or sample is added

What I did was load the JSON API whoami endpoint in one tab, noticed the boolean on one of the switches, opened the Django admin and toggled one of the switches, back to the JSON API whoami endpoint and refresh. The boolean hadn't changed under `switches`. 

If we move the `Switches.get_all()` into the view function it would read from the cache every time. So if we can just trigger a cache purge (from the actions in Django admin) then I think this "staleness problem" would go away.

Reply:
Thanks for testing that. Sounds like I badly misundersood how waffle.models.Switch works. I assumed that because is_active() is a method it would be checking the current setting when I called it and not just returning the value when I called get_all().

It really seems like overkill to mess with the cache and the admin interface. I'll just move the get_all() calls into the function itself. As I said, I'm just duplicating code that is already called on every single request, so I don't believe there is a perf issue we need to address.

Reply:
> just duplicating code that is already called on every single request 

according to New Relic rendering a document (i.e. the jinja dance) is, by 69%, the most time consuming function. Within, it averages 24.1 Redis get requests!! Making up ~20% of those 69%. I bet waffle flag fetching stuff is a good chunk of that. Some other day we can review that. Or just not and rewrite stuff with React and Stumptown.